### PR TITLE
Refactor: GLTF 2.0 Loader

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0 has all passed
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
Refactor.
### What is the current behavior? (You can also link to an open issue here)
Code is not modular enough and it is not easy to deal with bug and extensions.
### What is the new behavior (if this is a feature change)?
- [x] Refactor gltf 2.0 loader
    - [x] entry: GLTFParser
    - [x] Parsers:
       - [x]  BufferParser
       - [x] Validator
       - [x] TextureParser
       - [x] MaterialParser
       - [x] MeshParser
       - [x] SkinParser
       - [x] AnimationParser
       - [x] EntityParser
       - [x] SceneParser
    - [ ] Extensions
      - [x] KHR_draco_mesh_compression
      - [x] KHR_lights_punctual
      - [ ] KHR_materials_clearcoat (no support)
      - [ ] KHR_materials_ior  (no support)
      - [x] KHR_materials_pbrSpecularGlossiness
      - [ ] KHR_materials_sheen  (no support)
      - [ ] KHR_materials_transmission  (no support)
      - [x] KHR_materials_unlit
      - [x] KHR_materials_variants
      - [ ] KHR_materials_volume  (no support)
      - [x] KHR_mesh_quantization
      - [ ] KHR_texture_basisu  (no support, KTX2 needed)
      - [x] KHR_texture_transform
- [x] fix some bugs.
   - [x] add texture default name.
   - [x] use texture instead of image in `ITextureInfo.index`.
   - [x] process texture sampler.
   - [x] omit gltf camera. can use in `GLTFResource.cameras`
   - [x] `camera.fov` use radiance instead of angle.
   - [x] add count in no-indices  primitive.
   - [x] when `ISkin.skeleton` is undefined, joints should transforms resolve to scene root.
   - [x] support `accessor.sparse`
   - [x] aligment with 4-byte 
   - [x] use entity instead of entity name in `skin`
   - [x] material invalid name

### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
No.
### Other information:

### CODE REVIEW

File: packages/core/src/base/EngineObject.ts [hunk 1]
```diff
@@ -15,6 +15,13 @@ export abstract class EngineObject {
   @ignoreClone
   protected _engine: Engine;
 
+  /**
+   * Get the engine which the object belongs.
+   */
+  get engine(): Engine {
+    return this._engine;
+  }
+
```
Comment (GuoLei1990) [inline packages/core/src/base/EngineObject.ts:24]: If you add engine roperty getter here, you should delete `Scene`, `Entity`, `Buffer`’s engine property getter!

File: packages/loader/src/gltf/schema.ts [hunk 2]
```diff
@@ -623,71 +595,63 @@ export interface GlTf {
...
   extras?: any;
-  [k: string]: any;
-}
-
-export interface LoadedGLTFResource {
-  buffers?: ArrayBuffer[];
-  textures?: Texture2D[];
-  shaders?: string[];
-  gltf?: GlTf;
 }
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/schema.ts:657]: 1.The file name should `Schema.ts`, not `schema.ts`
2. You can write the interfaces with specification: https://github.com/KhronosGroup/glTF/tree/master/specification/2.0/schema , for example use `extension` to write the file same as official json.

File: packages/loader/src/gltf/GLTFResource.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { AnimationClip, BufferMesh, Camera, EngineObject, Entity, Material, Skin, Texture2D } from "@oasis-engine/core";
+import { IGLTF } from "./schema";
+
+export class GLTFResource extends EngineObject {
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/GLTFResource.ts:4]: Write comments for this class,this is a common class used by users and is very important.

Comment (zhuxudong) [reply packages/loader/src/gltf/GLTFResource.ts:4]: done

File: packages/loader/src/gltf/schema.ts (hunk 2 above)
Comment (zhuxudong) [reply packages/loader/src/gltf/schema.ts:657]: done

File: packages/core/src/base/EngineObject.ts (hunk 1 above)
Comment (zhuxudong) [reply packages/core/src/base/EngineObject.ts:24]: done

File: packages/core/src/Component.ts [hunk 4]
```diff
@@ -60,13 +60,6 @@ export abstract class Component extends EngineObject {
     return this._entity.scene;
   }
 
-  /**
-   * The engine which the component's entity belongs to.
-   */
-  get engine(): Engine {
-    return this._entity.engine;
-  }
-
```
Comment (GuoLei1990) [inline packages/core/src/Component.ts:69]: Revert, this not extends `EngineObject`, you can't delete.

File: packages/loader/src/gltf/Util.ts [hunk 5]
```diff
@@ -211,3 +249,73 @@ export function parseRelativeUrl(baseUrl: string, relativeUrl: string): string {
...
+    buffers.push(buffer);
+
+    byteOffset += chunkLength + 2 * UINT32_LENGTH;
+  }
+
+  return {
+    gltf,
+    buffers
+  };
+}
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/Util.ts:321]: Use class to warp this.

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**